File: packages/draftfirst/src/pdfsignal.ts
/**
 * PDF round-trip signal — how a Draft First PDF carries its own source home.
 *
 * Parsing arbitrary PDF text back into a screenplay is lossy and would take
 * a dependency we refuse to carry. But a PDF WE exported never needs
 * parsing: its Info dictionary can legally hold the complete Fountain source
 * as a hex string in /Keywords, a field every viewer and print pipeline
 * preserves. Export stamps it; import scans for it. Perfect fidelity, zero
 * dependencies, and foreign PDFs simply have no signal — a clean refusal.
 *
 * The payload is versioned. A future format bumps the version so tomorrow's
 * reader can tell today's files apart — and today's reader says "I can't"
 * instead of reading garbage.
 *
 * Encoding note: the payload hex-encodes UTF-8 BYTES, not UTF-16 code
 * units — curly quotes, em dashes, and every non-Latin script survive
 * intact. (Hex-encoding code units, the tempting shortcut, corrupts beyond
 * the BMP and mangles nothing less than the writer's own punctuation.)
 */

import { decodeUtf8, encodeUtf8 } from './platform.js';

export const PDF_MARKER_PREFIX = 'DRAFT_FIRST_FOUNTAIN';
export const PDF_MARKER_VERSION = '1';

/** The hex string to stamp into /Keywords at export. */
export function encodePdfPayload(fountain: string): string {
	const bytes = encodeUtf8(`${PDF_MARKER_PREFIX}:${PDF_MARKER_VERSION}\n${fountain}`);
	let hex = '';
	for (let i = 0; i < bytes.length; i++) hex += bytes[i]!.toString(16).padStart(2, '0');
	return hex;
}

const KEYWORD_FIELD = '/Keywords';
const HEX_DIGIT = /^[0-9a-fA-F]$/;

/** The raw bytes of a `/Keywords <hex>` value starting at `from`, or null. */
function readKeywordsHex(source: Uint8Array, from: number): string | null {
	let at = from + KEYWORD_FIELD.length;
	while (at < source.length && (source[at] === 0x20 || source[at] === 0x09 || source[at] === 0x0a || source[at] === 0x0d)) {
		at++;
	}
	if (source[at] !== 0x3c /* '<' */) return null;
	if (source[at + 1] === 0x3c /* '<<' — a dictionary, not a hex string */) return null;
	at++;
	let hex = '';
	while (at < source.length && source[at] !== 0x3e /* '>' */) {
		const ch = String.fromCharCode(source[at]!);
		if (!HEX_DIGIT.test(ch)) return null;
		hex += ch;
		at++;
	}
	if (at >= source.length || hex.length === 0 || hex.length % 2 !== 0) return null;
	return hex;
}

function hexToBytes(hex: string): Uint8Array {
	const bytes = new Uint8Array(hex.length / 2);
	for (let i = 0; i < bytes.length; i++) bytes[i] = Number.parseInt(hex.slice(i * 2, i * 2 + 2), 16);
	return bytes;
}

/**
 * The Fountain source embedded in a Draft First PDF, or null when the file
 * carries no valid signal — a foreign PDF, an older export, or a newer
 * format version this build does not understand.
 */
export function extractPdfPayload(source: Uint8Array): string | null {
	const needle = encodeUtf8(KEYWORD_FIELD);
	outer: for (let i = 0; i + needle.length <= source.length; i++) {
		for (let n = 0; n < needle.length; n++) {
			if (source[i + n] !== needle[n]) continue outer;
		}
		const hex = readKeywordsHex(source, i);
		if (hex === null) continue;
		const payload = decodeUtf8(hexToBytes(hex));
		const separator = payload.indexOf('\n');
		if (separator < 0) continue;
		const header = payload.slice(0, separator);
		const [prefix, version] = header.split(':');
		if (prefix !== PDF_MARKER_PREFIX || version !== PDF_MARKER_VERSION) continue;
		return payload.slice(separator + 1);
	}
	return null;
}

File: ios/EDraftCore/Sources/EDraftCore/PdfSignal.swift
import Foundation

/// How an eDraft PDF carries its own Fountain source home.
///
/// Parsing arbitrary PDF text back into a screenplay is lossy. A PDF we
/// exported never needs parsing: its bytes carry the complete Fountain
/// source as a hex string in `/Keywords`, a field every viewer and print
/// pipeline preserves. Export stamps it; import scans for it. Perfect
/// fidelity, zero PDF-graph dependencies, and a foreign PDF simply has
/// no signal — a clean refusal.
///
/// The payload hex-encodes UTF-8 *bytes*, not UTF-16 code units — curly
/// quotes, em dashes, and every non-Latin script survive intact. Mirrors
/// `packages/edraft/src/pdfsignal.ts` byte for byte.
///
/// Two `/Keywords` spellings are read: a PDF hex string `<…>` (the web
/// exporter) and a PDF literal `(…)` (Core Graphics). Both live in the
/// Info dictionary. Bytes after `%%EOF` are not a field and are not how
/// we write.
public nonisolated enum PdfSignal {

    public static let markerPrefix = "EDRAFT_FOUNTAIN"
    public static let markerVersion = "1"

    /// Prefixes written before the eDraft rename. We read them forever and
    /// never write them again — those files are writers' backups.
    public static let legacyMarkerPrefixes = ["EDRAFT_FOUNTAIN"]

    /// The hex string to stamp into `/Keywords` at export.
    public static func encode(_ fountain: String) -> String {
        hex(of: Array("\(markerPrefix):\(markerVersion)\n\(fountain)".utf8))
    }

    /// The Fountain source embedded in an eDraft PDF, or `nil` when the
    /// file carries no valid signal — a foreign PDF, a truncated write,
    /// or a newer format version this build does not understand.
    public static func extract(from pdf: Data) -> String? {
        let source = pdf
        let needle = Array("/Keywords".utf8)
        var i = 0
        outer: while i + needle.count <= source.count {
            for n in 0..<needle.count {
                if source[i + n] != needle[n] {
                    i += 1
                    continue outer
                }
            }
            if let hex = readKeywordsValue(source, from: i),
               let payload = decodePayload(hex) {
                return payload
            }
            i += 1
        }
        return nil
    }

    // MARK: - Internals

    private static func isKnownPrefix(_ prefix: String) -> Bool {
        prefix == markerPrefix || legacyMarkerPrefixes.contains(prefix)
    }

    /// Hex digits of a `/Keywords` value. Accepts a PDF hex string `<…>`
    /// (the web exporter) and a PDF literal `(…)` (Core Graphics
    /// `kCGPDFContextKeywords`). The payload is hex either way, so both
    /// sit in the Info dictionary and survive a viewer re-save.
    private static func readKeywordsValue(_ source: Data, from: Int) -> String? {
        var at = from + "/Keywords".utf8.count
        while at < source.count {
            let b = source[at]
            if b == 0x20 || b == 0x09 || b == 0x0a || b == 0x0d {
                at += 1
                continue
            }
            break
        }
        guard at < source.count else { return nil }
        let closer: UInt8
        if source[at] == 0x3c {
            // `<<` is a dictionary, not a hex string.
            if at + 1 < source.count, source[at + 1] == 0x3c { return nil }
            closer = 0x3e
        } else if source[at] == 0x28 {
            closer = 0x29
        } else {
            return nil
        }
        at += 1
        var hex = ""
        while at < source.count, source[at] != closer {
            let ch = Character(UnicodeScalar(source[at]))
            guard ch.isHexDigit else { return nil }
            hex.append(ch)
            at += 1
        }
        guard at < source.count, !hex.isEmpty, hex.count.isMultiple(of: 2) else { return nil }
        return hex
    }

    private static func decodePayload(_ hex: String) -> String? {
        guard let bytes = bytes(fromHex: hex),
              let payload = String(bytes: bytes, encoding: .utf8)
        else { return nil }
        guard let separator = payload.firstIndex(of: "\n") else { return nil }
        let header = payload[..<separator]
        let parts = header.split(separator: ":", maxSplits: 1, omittingEmptySubsequences: false)
        guard parts.count == 2 else { return nil }
        let prefix = String(parts[0])
        let version = String(parts[1])
        guard isKnownPrefix(prefix), version == markerVersion else { return nil }
        return String(payload[payload.index(after: separator)...])
    }

    private static func hex(of bytes: [UInt8]) -> String {
        bytes.map { String(format: "%02x", $0) }.joined()
    }

    private static func bytes(fromHex hex: String) -> [UInt8]? {
        var bytes: [UInt8] = []
        bytes.reserveCapacity(hex.count / 2)
        var index = hex.startIndex
        while index < hex.endIndex {
            let next = hex.index(index, offsetBy: 2)
            guard let byte = UInt8(hex[index..<next], radix: 16) else { return nil }
            bytes.append(byte)
            index = next
        }
        return bytes
    }
}
